File: backend-elixir/test/support/test_token_helper.ex
defmodule BackendElixir.TestSupport.TestTokenHelper do
  @moduledoc "backend-java/backend-kotlin/backend-pythonのtest_token_helper相当"

  def make_hmac_token(secret, iss, aud, sub, exp_offset_secs, alg \\ "HS256") do
    signer = Joken.Signer.create(alg, secret)
    claims = %{"sub" => sub, "iss" => iss, "aud" => aud, "exp" => System.system_time(:second) + exp_offset_secs}
    {:ok, token, _claims} = Joken.encode_and_sign(claims, signer)
    token
  end

  @doc """
  "alg":"none"は署名の要らない自己主張トークンになるため、必ず拒否されなければならない。
  Jokenは通常algをsigner経由で扱うため、"none"は手動で組み立てる
  (backend-java/backend-kotlin/backend-pythonの同名ヘルパーと同じ意図)
  """
  def make_alg_none_token(iss, aud, sub, exp_offset_secs) do
    header = %{"alg" => "none", "typ" => "JWT"}
    claims = %{"sub" => sub, "iss" => iss, "aud" => aud, "exp" => System.system_time(:second) + exp_offset_secs}
    header_b64 = b64url(Jason.encode!(header))
    payload_b64 = b64url(Jason.encode!(claims))
    "#{header_b64}.#{payload_b64}."
  end

  def generate_rsa_key_pair do
    :public_key.generate_key({:rsa, 2048, 65537})
  end

  def make_rsa_token(private_key, kid, iss, aud, sub, exp_offset_secs, alg \\ "RS256") do
    jwk = private_jwk(private_key)
    signer = Joken.Signer.create(alg, jwk, %{"kid" => kid})
    claims = %{"sub" => sub, "iss" => iss, "aud" => aud, "exp" => System.system_time(:second) + exp_offset_secs}
    {:ok, token, _claims} = Joken.encode_and_sign(claims, signer)
    token
  end

  @doc "外部公開API向け(azp=Client Credentials Grantのクライアントid)のRS256トークンを組み立てる"
  def make_rsa_token_with_azp(private_key, kid, iss, aud, sub, azp, exp_offset_secs, alg \\ "RS256") do
    jwk = private_jwk(private_key)
    signer = Joken.Signer.create(alg, jwk, %{"kid" => kid})

    claims = %{
      "sub" => sub,
      "iss" => iss,
      "aud" => aud,
      "azp" => azp,
      "exp" => System.system_time(:second) + exp_offset_secs
    }

    {:ok, token, _claims} = Joken.encode_and_sign(claims, signer)
    token
  end

  def public_jwk_fields(private_key) do
    {:RSAPrivateKey, _, modulus, exponent, _, _, _, _, _, _, _} = private_key
    %{"n" => b64url_uint(modulus), "e" => b64url_uint(exponent)}
  end

  defp private_jwk(private_key) do
    {:RSAPrivateKey, _, modulus, pub_exp, priv_exp, _, _, _, _, _, _} = private_key

    %{
      "kty" => "RSA",
      "n" => b64url_uint(modulus),
      "e" => b64url_uint(pub_exp),
      "d" => b64url_uint(priv_exp)
    }
  end

  defp b64url_uint(value) do
    value |> :binary.encode_unsigned() |> b64url()
  end

  defp b64url(data), do: Base.url_encode64(data, padding: false)
end

File: backend-haskell/test/Support/TestTokenHelper.hs
-- | backend-java/backend-kotlin/backend-python/backend-elixirのtest_token_helper相当
module TestTokenHelper
  ( makeHmacToken
  , makeHmacTokenWithAzp
  , makeAlgNoneToken
  , generateRsaKeyPair
  , makeRsaToken
  , makeRsaTokenWithAzp
  , publicJwkFields
  ) where

import Crypto.JOSE (runJOSE)
import qualified Crypto.JOSE.JWK as JWK
import qualified Crypto.JWT as JWT
import Control.Lens ((&), (.~), (?~), preview)
import qualified Crypto.PubKey.RSA as RSA
import qualified Data.Aeson as Aeson
import qualified Data.ByteString.Base64.URL as B64URL
import qualified Data.ByteString.Char8 as BC
import qualified Data.ByteString.Lazy as BL
import qualified Data.Map as M
import Data.Text (Text)
import qualified Data.Text as T
import qualified Data.Text.Encoding as TE
import Data.Time (addUTCTime, getCurrentTime)
import Data.Time.Clock.POSIX (utcTimeToPOSIXSeconds)

makeHmacToken :: Text -> Text -> Text -> Text -> Integer -> IO Text
makeHmacToken secret iss aud sub expOffsetSecs = signWith (JWK.fromOctets (TE.encodeUtf8 secret)) JWT.HS256 Nothing iss aud sub expOffsetSecs Nothing

makeRsaToken :: RSA.PrivateKey -> Text -> Text -> Text -> Text -> Integer -> IO Text
makeRsaToken privateKey kid iss aud sub expOffsetSecs = do
  let jwk = privateJwk privateKey kid
  signWith jwk JWT.RS256 (Just kid) iss aud sub expOffsetSecs Nothing

-- | 外部公開APIのRequireExternalClientAuth相当のテスト専用: azp(トークンを取得した
-- OAuth2クライアントのclient_id)を私的クレームとして追加したローカルHMACトークンを作る
makeHmacTokenWithAzp :: Text -> Text -> Text -> Text -> Integer -> Text -> IO Text
makeHmacTokenWithAzp secret iss aud sub expOffsetSecs azp =
  signWith (JWK.fromOctets (TE.encodeUtf8 secret)) JWT.HS256 Nothing iss aud sub expOffsetSecs (Just azp)

-- | 同上、RS256(Keycloak発行相当)版
makeRsaTokenWithAzp :: RSA.PrivateKey -> Text -> Text -> Text -> Text -> Integer -> Text -> IO Text
makeRsaTokenWithAzp privateKey kid iss aud sub expOffsetSecs azp = do
  let jwk = privateJwk privateKey kid
  signWith jwk JWT.RS256 (Just kid) iss aud sub expOffsetSecs (Just azp)

signWith :: JWK.JWK -> JWT.Alg -> Maybe Text -> Text -> Text -> Text -> Integer -> Maybe Text -> IO Text
signWith jwk alg mKid iss aud sub expOffsetSecs mAzp = do
  now <- getCurrentTime
  let expAt = addUTCTime (fromIntegral expOffsetSecs) now
      baseClaims =
        JWT.emptyClaimsSet
          & JWT.claimIss ?~ fromStringOrUri iss
          & JWT.claimSub ?~ fromStringOrUri sub
          & JWT.claimAud ?~ JWT.Audience [fromStringOrUri aud]
          & JWT.claimExp ?~ JWT.NumericDate expAt
      -- azpは標準クレームではなくprivate claim(未登録クレーム)のため、
      -- unregisteredClaimsへ直接差し込む(HmacVerifier/JwksVerifierのazpOfと対になる)
      claims = case mAzp of
        Nothing -> baseClaims
        Just azp -> baseClaims & JWT.unregisteredClaims .~ M.fromList [("azp", Aeson.String azp)]
      header =
        JWT.newJWSHeaderProtected alg
          & JWT.kid .~ fmap (JWT.newHeaderParamProtected) mKid
  result <- runJOSE (JWT.signClaims jwk header claims)
  case result of
    Left err -> error ("TestTokenHelper: failed to sign token: " ++ show (err :: JWT.JWTError))
    Right signedJwt -> pure (TE.decodeUtf8 (BL.toStrict (JWT.encodeCompact signedJwt)))

fromStringOrUri :: Text -> JWT.StringOrURI
fromStringOrUri t = case preview JWT.stringOrUri t of
  Just v -> v
  Nothing -> error ("TestTokenHelper: invalid StringOrURI: " ++ T.unpack t)

-- | "alg":"none"は署名の要らない自己主張トークンになるため、必ず拒否されなければならない。
-- joseは通常algをsigner経由で扱うため、"none"は手動で組み立てる
makeAlgNoneToken :: Text -> Text -> Text -> Integer -> IO Text
makeAlgNoneToken iss aud sub expOffsetSecs = do
  now <- getCurrentTime
  let expAt = round (realToFrac (utcTimeToPOSIXSeconds (addUTCTime (fromIntegral expOffsetSecs) now)) :: Double) :: Integer
      header = Aeson.object ["alg" Aeson..= ("none" :: Text), "typ" Aeson..= ("JWT" :: Text)]
      payload = Aeson.object ["sub" Aeson..= sub, "iss" Aeson..= iss, "aud" Aeson..= aud, "exp" Aeson..= expAt]
      headerB64 = b64urlEncode (BL.toStrict (Aeson.encode header))
      payloadB64 = b64urlEncode (BL.toStrict (Aeson.encode payload))
  pure (headerB64 <> "." <> payloadB64 <> ".")

b64urlEncode :: BC.ByteString -> Text
b64urlEncode = TE.decodeUtf8 . BC.dropWhileEnd (== '=') . B64URL.encode

generateRsaKeyPair :: IO (RSA.PublicKey, RSA.PrivateKey)
generateRsaKeyPair = RSA.generate 256 65537

privateJwk :: RSA.PrivateKey -> Text -> JWK.JWK
privateJwk priv kid =
  JWK.fromKeyMaterial (JWK.RSAKeyMaterial (JWK.toRSAKeyParameters priv))
    & JWK.jwkKid ?~ kid

-- | JWKSモックサーバーへ登録する公開鍵の"n"/"e"フィールド(base64url、パディング無し)
publicJwkFields :: RSA.PublicKey -> (Text, Text)
publicJwkFields pub = (b64urlUint (RSA.public_n pub), b64urlUint (RSA.public_e pub))

b64urlUint :: Integer -> Text
b64urlUint n = b64urlEncode (integerToBytes n)

integerToBytes :: Integer -> BC.ByteString
integerToBytes n = BC.pack (map toEnum (go n []))
  where
    go 0 acc = if null acc then [0] else acc
    go x acc = go (x `div` 256) (fromIntegral (x `mod` 256) : acc)
